### PR TITLE
Update for v2, streamline API some

### LINKED ISSUE #65: Add support to render sync
I am interested on running this library embedded in a .NET application using [JavaScriptEngineSwitcher](https://github.com/Taritsyn/JavaScriptEngineSwitcher)

Getting back a string when executing render will be ideal.

 Is there any way to be able to:
`var html = render(m('div', 'Hello'))`
`//<div>Hello</div>`

I don't care on losing the async initialization.

### LINKED ISSUE #85: Adapt docs to async/await

### LINKED ISSUE #90: Mithril's `oninit` is synchronous
Mithril core's `oninit` is synchronous, yet this for some reason awaits it. Shouldn't it be sync as well, to match the rest of the API including core's `m.render` being itself 100% sync?

It'd also resolve #65 as a side effect.

### PR DESCRIPTION
**Pull request by isiahmeadows**
**Edit:** Update reference to file

Note before merging: you *will* need to bump this in a major version before you release it. There are a few breaking changes in the API to streamline it a little.

Everything passes tests, but it *really* could use more tests. Do note that the `tests.js` diff, although it may look massive, is mostly just indentation changes, `async`/`await` removal (along with the corresponding `o.async` local helper), ES6-ifying methods and functions, and other relatively procedural refactoring, with only a few tests changed beyond that. Most of the interesting bits are in `index.js`, and you might find it easier to just look at the [new file itself](https://github.com/isiahmeadows/mithril-node-render/blob/v2/index.js) and comparing it with the old file in a separate tab/window to see the changes instead of reading the fairly verbose diff.

-----

- Make everything synchronous as is the case in core
- `escapeAttributeValue` -> `escapeAttribute`
- `escapeString` -> `escapeText`
- Add `xml` for actual XML, limit `strict` to void HTML elements
- Use ospec from npm
- Use Mithril's `mithril/render/vnode` to help simplify the module
- Align behavior better with Mithril itself

Fixes #90
Fixes #65
Fixes #85

Basically had to half rewrite the module. I also included a few various minor perf optimizations, including but not limited to:

- I switched how strings were built, to leave a hint to the engine that it should use a string builder representation rather than just a bunch of cons strings. (It's a thing template engines also should be aware of.)
- The void tag test is now a regexp test, so it's O(n) instead of O(n^2).
- A merged replacer function is *slightly* faster than multiple string replacements when the replacer function is shared across all calls, because the regexp is relatively simple and it only needs to iterate the string once. (Strings aren't arrays, so iterating them in general isn't cache-friendly.)
- I included a few optimization tricks used also in Mithril on the create side. The new source somewhat mirrors this.
- Changed everything to work in a single closure to try to reduce stack usage just a hair. (I didn't do anything crazy like manually code tail calls or anything.)

So it should be noticeably faster now, too.

And as a drive-by, I corrected the ospec module to use a few more tests where it was previously using assertion messages. (Makes failures a lot easier to debug when it doesn't prematurely abort.) I'll leave improving the test battery for later for you or whoever later feels the urge to.

### CODE REVIEW

**Comment by StephanHoyer** [comment]
thanks @isiahmeadows. Will look today

**Comment by StephanHoyer** [comment]
Why did you remove async behaviour? I think it's essential part of the module, because otherwise we have a problem waiting for db fetches and stuff. Sure this might be done before starting rendering but it simplifies a lot for me that I can simply return an unresolved promise from a `oninit` to delay rendering.

**Comment by isiahmeadows** [comment]
@StephanHoyer Updated per our discussion on Mithril's Gitter to use a `waitFor` callback on async renders. I've also updated the README in light of this.

I ended up switching to a generator for the short term, just to get something simple and working. In the future, I'd like to see about maybe moving to a state machine-based approach for a little more performance, but I don't see this as likely any time before v3.

File: README.md [hunk 1]
```diff
@@ -59,6 +63,28 @@ render(myAsyncComponent).then(function (html) {
...
+      node.state.foo = 'bar'
+      resolve()
+    })
+  },
+  view: function (node) {
+    return m('div', node.state.foo)
+  }
+}
+
+var html = render.sync(myAsyncComponent)
```
Comment (StephanHoyer) [inline README.md]: does this really work? Don't get how you where able to convert a async call to a sync one.

Comment (StephanHoyer) [reply README.md]: will try that, if I found time at the weekend

**Comment by StephanHoyer** [comment]
Thanks a lot. Awesome solution. Just published 3.0

**Pull request merged**